programs/bpf_loader/src/upgradeable/processor.rs

```diff
@@ -0,0 +1,102 @@
+use {
+    solana_program_runtime::{ic_logger_msg, invoke_context::InvokeContext},
+    solana_sdk::{
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        instruction::InstructionError,
+        system_instruction::{self, MAX_PERMITTED_DATA_LENGTH},
+    },
+};
+
+pub(crate) fn process_extend_program_data(
+    invoke_context: &mut InvokeContext,
+    additional_bytes: usize,
+) -> Result<(), InstructionError> {
+    let log_collector = invoke_context.get_log_collector();
+    let transaction_context = &invoke_context.transaction_context;
+    let instruction_context = transaction_context.get_current_instruction_context()?;
+    let program_id = instruction_context.get_last_program_key(transaction_context)?;
+
+    if additional_bytes == 0 {
+        ic_logger_msg!(log_collector, "Additional bytes must be greater than 0");
+        return Err(InstructionError::InvalidInstructionData);
+    }
+
+    const PROGRAM_DATA_ACCOUNT_INDEX: usize = 0;
+    #[allow(dead_code)]
+    const SYSTEM_PROGRAM_ACCOUNT_INDEX: usize = 1;
+    const OPTIONAL_PAYER_ACCOUNT_INDEX: usize = 2;
+
+    let programdata_account = instruction_context
+        .try_borrow_instruction_account(transaction_context, PROGRAM_DATA_ACCOUNT_INDEX)?;
+    let programdata_key = *programdata_account.get_key();
+
```

Comment:
What about `is_executable()` and `is_writable()`?

Reply:
Those violations should be caught by the runtime or your new changes in `BorrowedAccount` (https://github.com/solana-labs/solana/pull/25899) so it's redundant to check here

Reply:
I am not concerned about the rules not being applied, but that there is no helpful error message in that case. See: https://github.com/solana-labs/solana/blob/d18e2f929e974424d697b55fd3fb87697885acae/programs/bpf_loader/src/lib.rs#L740

Reply:
Ah ok, the linked checks are critical in the `Upgrade` instruction, they aren't for helpful error messages. In this case it's not possible for a programdata account to become executable. I'll add a writable check for now but we can remove it after invalid mutations are detected by https://github.com/solana-labs/solana/pull/25899